assets/src/edit-story/app/history/historyProvider.js

```diff
@@ -40,23 +40,29 @@ function HistoryProvider({ children, size }) {
     versionNumber,
   } = useHistoryReducer(size);
 
+  const [hasNewChanges, setHasNewChanges] = useState(false);
   const setPreventUnload = usePreventWindowUnload();
 
   useEffect(() => {
-    setPreventUnload('history', versionNumber - 1 > 0);
-
+    setPreventUnload('history', hasNewChanges);
     return () => setPreventUnload('history', false);
-  }, [setPreventUnload, versionNumber]);
+  }, [setPreventUnload, hasNewChanges]);
+
+  useEffect(() => {
+    setHasNewChanges(versionNumber - 1 > 0);
+  }, [versionNumber]);
 
   const state = {
     state: {
       replayState,
+      hasNewChanges,
       canUndo: offset < historyLength - 1,
       canRedo: offset > 0,
     },
     actions: {
       appendToHistory,
       clearHistory,
+      setHasNewChanges,
```

Comment:
IMHO the best way to export this is as a no-argument `resetNewChanges()` method or something like this. I.e. only limit it to update the `hasNewChanges` to `false`. I think we really don't want to allow others to call it with `setHasNewChanges(true)` except for this one provider.

Reply:
Makes sense, will do. 

Will have to look into this in more detail anyway: apparently, the logic of using only the history version is not really accurate -- it doesn't consider the combination of offset + the history version, e.g. if the user saves and then clicks "Undo", it's not considered a change -- this means that closing the tab won't be prevented either.

Reply:
I think logically we have to make undo/redo to always increment the version in the history reducer. If that doesn't happen right now, we can fix in a separate pull request (or this one if it's a small fix).

Reply:
Yes, also, here the new changes have to be tracked against the last saved state, just tracking the change isn't enough. E.g. if the user saves, then clicks undo, it should set having new changes to `true`, and then if the user clicks redo, it should set the new changes to `false` again.

Reply:
Hmm, we should probably fix this within the current issue since the Save Draft button gets disabled if there are no new changes -- otherwise, it would prevent saving the story.

Will take a look tomorrow.

Reply:
Actually, added the logic already now, PTAL.